cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll

```diff
@@ -832,10 +893,19 @@ module FlowVar_internal {
     )
   }
 
+  Call getAPointerWrapperAccess(Variable pointer) {
+    pointer.getUnspecifiedType() instanceof PointerWrapper and
+    [result.getQualifier(), result.getAnArgument()] = pointer.getAnAccess()
+  }
```

Comment:
Yeah, good point. The name is a result of me following the pattern from `getAnIteratorAccess` which has a more complicated job to do.

The predicate is used by the char pred of `SmartPointerPartialDefinition`. In the case of `operator->` on a smart pointer the `valueToUpdate` predicate on line 254:
```codeql
valueToUpdate(convertedInner, this.getFullyConverted(), node) and
```
follows the flow from `this.getFullyConverted()` down to the call to `operator->`, and the `getAPointerWrapperAccess` predicate then related that call to the variable being updated.

The problem with this predicate is really that it serves two purposes depending on which disjunction of the char pred we're in:
1. In the `valueToUpdate` case (which covers the case of `smart_ptr->x`) we want to return the qualifier, and
2. In the copy constructor case (which covers the case of passing a smart pointer by value to a function) we want to return the 0'th argument since copy constructors don't have a qualifier).

After writing down this explanation I realize that `getAPointerWrapperAccess` can be specialized more than what it's currently doing. I've fixed this in 0a6aef71a2f957386f895d594f227a0c67c387be (which hopefully should also make the goal of the predicate more clear).